lms/djangoapps/experiments/flags.py

```diff
@@ -162,16 +180,28 @@ def get_bucket(self, course_key=None, track=True):
             if not self._is_enrollment_inside_date_bounds(values, user, course_key):
                 return self._cache_bucket(experiment_name, 0)
 
-        bucket = stable_bucketing_hash_group(experiment_name, self.num_buckets, user.username)
-
-        # Now check if the user is forced into a particular bucket, using our subordinate bucket flags
+        # Determine the user's bucket.
+        # First check if forced into a particular bucket, using our subordinate bucket flags.
+        # If not, calculate their default bucket using a consistent hash function.
         for i, bucket_flag in enumerate(self.bucket_flags):
-            if bucket_flag.is_enabled(course_key):
+            forced_into_bucket = (
+                bucket_flag.is_enabled(course_key) if course_key
+                else bucket_flag.is_enabled_without_course_context()
```

Comment:
I believe my original code for this class called `_is_enabled` before someone (me?) decided to avoid calling another class's private methods. So good fix, just calling `is_enabled` seems like it was just patently wrong.

This construction is so painful though. Do we really need `CourseWaffleFlag.is_enabled` to reject a `None` course? It seems fine to allow it as an optional flag and only run the `assert` if it's not `None`.

But sure, we can do this logic for now, if you don't want to refactor that.

Reply:
Yes, the whole thing seems pretty unidiomatic. I'll see how hard it'd be to refactor it; if it's easy, I'll throw another commit on this PR.